addons/project/models/project_task.py

```diff
@@ -722,41 +722,74 @@ def copy_data(self, default=None):
         not_project_user = not self.env.user.has_group('project.group_project_user')
         if not_project_user:
             vals_list = [{k: v for k, v in vals.items() if k in self.SELF_READABLE_FIELDS} for vals in vals_list]
+
+        milestone_mapping = self.env.context.get('milestone_mapping', {})
         for task, vals in zip(self, vals_list):
-            has_default_name = bool(default.get('name', ''))
-            if not has_default_name:
-                vals['name'] = _("%s (copy)", task.name)
-                vals['child_ids'] = [child.copy({'name': _("%s (copy)", child.name)}).id for child in task.child_ids]
-            else:
-                vals['child_ids'] = [child.copy({'name': child.name}).id for child in task.child_ids]
+
+            vals['stage_id'] = task.stage_id.id
+            vals['name'] = task.name if self.env.context.get('copy_project') else _("%s (copy)", task.name)
             if task.recurrence_id:
                 vals['recurrence_id'] = task.recurrence_id.copy().id
+            if task.allow_milestones:
+                vals['milestone_id'] = milestone_mapping.get(vals['milestone_id'], vals['milestone_id'])
+            if task.child_ids:
+                default = {
+                    'depend_on_ids': False,
+                    'dependent_ids': False,
+                    'child_ids': False,
+                    'parent_id': False,
+                }
+                vals['child_ids'] = [Command.create(child_id.copy_data(default)[0]) for child_id in task.child_ids]
         return vals_list
 
+    def create_task_mapping(self, copied_task):
```

Comment:
```suggestion
    def create_task_mapping(self, copied_tasks):
```
Sorry for not making it clear the first time. This is what I meant on L755.